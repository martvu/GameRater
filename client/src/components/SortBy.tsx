import {
  Select,
  SelectContent,
  SelectGroup,
  SelectItem,
  SelectTrigger,
  SelectValue,
} from '@/components/ui/select';

type SortByProps = {
  onSort: (field: string, order: string) => void;
};

export default function SortBy({ onSort }: SortByProps) {
  function handleSortSelection(sortBy: string) {
    if (sortBy === 'release date asc') {
      onSort('first_release_date', 'asc');
    }
    if (sortBy === 'release date desc') {
      onSort('first_release_date', 'desc');
    }
    if (sortBy === 'name') {
      onSort('name', 'asc');
    }
    if (sortBy === 'metascore') {
      onSort('aggregated_rating', 'desc');
    }
    if (sortBy === 'user_rating') {
      onSort('user_rating', 'desc');
    }
    console.log('Hello sort');
    // Store the selected value in localStorage
    localStorage.setItem('selectedSortBy', sortBy);
  }

  return (
    <Select onValueChange={sortBy => handleSortSelection(sortBy)}>
      <SelectTrigger className="w-[150px]">
        <SelectValue placeholder="Sort By" />
      </SelectTrigger>
      <SelectContent>
        <SelectGroup>
          <SelectItem className="cursor-pointer" value="release date asc">
            Release Date (ASC)
          </SelectItem>
          <SelectItem className="cursor-pointer" value="release date desc">
            Release Date (DESC)
          </SelectItem>
          <SelectItem className="cursor-pointer" value="name">
            Title
          </SelectItem>
          <SelectItem className="cursor-pointer" value="metascore">
            Metascore
          </SelectItem>
          <SelectItem className="cursor-pointer" value="user_rating">
            User Rating
          </SelectItem>
        </SelectGroup>
      </SelectContent>
    </Select>
  );
}
