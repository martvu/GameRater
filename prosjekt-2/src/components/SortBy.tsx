import {
  Select,
  SelectContent,
  SelectGroup,
  SelectItem,
  SelectLabel,
  SelectTrigger,
  SelectValue,
} from '@/components/ui/select';

export default function SelectDemo() {
  return (
    <Select>
      <SelectTrigger className="w-[150px]">
        <SelectValue placeholder="Sort By" />
      </SelectTrigger>
      <SelectContent>
        <SelectGroup>
          <SelectLabel>Sort By</SelectLabel>
          <SelectItem className="cursor-pointer"value="apple">Release Date</SelectItem>
          <SelectItem className="cursor-pointer"value="banana">Title</SelectItem>
          <SelectItem className="cursor-pointer"value="blueberry">Rating</SelectItem>
          <SelectItem className="cursor-pointer"value="grapes">Popularity</SelectItem>
        </SelectGroup>
      </SelectContent>
    </Select>
  );
}
