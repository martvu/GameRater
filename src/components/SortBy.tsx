import {
  Select,
  SelectContent,
  SelectGroup,
  SelectItem,
  SelectLabel,
  SelectTrigger,
  SelectValue,
} from '@/components/ui/select';

export default function SortBy() {
  return (
    <>
      <Select>
        <SelectTrigger className="w-[150px]">
          <SelectValue placeholder="Sort By" />
        </SelectTrigger>
        <SelectContent>
          <SelectGroup>
            <SelectLabel>Sort By</SelectLabel>
            <SelectItem className="cursor-pointer" value="release date">
              Release Date
            </SelectItem>
            <SelectItem className="cursor-pointer" value="title">
              Title
            </SelectItem>
            <SelectItem className="cursor-pointer" value="rating">
              Rating
            </SelectItem>
            <SelectItem className="cursor-pointer" value="popularity">
              Popularity
            </SelectItem>
          </SelectGroup>
        </SelectContent>
      </Select>
    </>
  );
}
