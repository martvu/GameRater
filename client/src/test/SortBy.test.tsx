import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { render, screen } from '@testing-library/react';
import SortBy from '@/components/SortBy';
import * as recoil from 'recoil';
import userEvent from '@testing-library/user-event';
import { RecoilRoot } from 'recoil';

/* Need to mock PointerEvent for the SortBy component tests to work 
because Radix UI uses PointerEvent */
class MockPointerEvent extends Event {
  button: number;
  ctrlKey: boolean;
  pointerType: string;

  constructor(type: string, props: PointerEventInit) {
    super(type, props);
    this.button = props.button || 0;
    this.ctrlKey = props.ctrlKey || false;
    this.pointerType = props.pointerType || 'mouse';
  }
}
//eslint-disable-next-line @typescript-eslint/no-explicit-any
window.PointerEvent = MockPointerEvent as any;
window.HTMLElement.prototype.scrollIntoView = vi.fn();
window.HTMLElement.prototype.releasePointerCapture = vi.fn();
window.HTMLElement.prototype.hasPointerCapture = vi.fn();

// Mock localStorage
type LocalStorageMock = {
  getItem: (key: string) => string | null;
  setItem: (key: string, value: string) => void;
  clear: () => void;
};

// Mock Recoil state
vi.mock('recoil', async () => {
  const actual: typeof recoil = await vi.importActual('recoil');
  return {
    ...actual, // Include all actual module exports
    useRecoilState: vi
      .fn<[], [unknown, (val: unknown) => void]>()
      .mockReturnValue([{}, vi.fn()]),
    useSetRecoilState: vi
      .fn<[], (val: unknown) => void>()
      .mockReturnValue(vi.fn()),
  };
});

const renderSortBy = () => {
  return render(
    <RecoilRoot>
      <SortBy />
    </RecoilRoot>
  );
};

describe('SortBy Component', () => {
  let localStorageMock: LocalStorageMock;
  beforeEach(() => {
    localStorageMock = {
      getItem: vi.fn(() => null),
      setItem: vi.fn(() => {}),
      clear: vi.fn(),
    };
    Object.defineProperty(window, 'localStorage', { value: localStorageMock });
  });

  afterEach(() => {
    localStorageMock.clear();
  });

  it('renders correctly', async () => {
    renderSortBy();
    expect(screen.getByText('Sort By')).toBeInTheDocument();
  });

  it('matches the snapshot', async () => {
    const { asFragment } = renderSortBy();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should open the select menu when clicked', async () => {
    renderSortBy();
    const trigger = screen.getByTestId('sort-by-select');
    await userEvent.click(trigger);
    expect(screen.getByText('Release Date Desc')).toBeInTheDocument();
  });

  it('should change the selected value when clicked', async () => {
    renderSortBy();
    const trigger = screen.getByTestId('sort-by-select');
    await userEvent.click(trigger);
    expect(trigger).toHaveAttribute('aria-expanded', 'true');
    await userEvent.click(screen.getByTestId('release-date-desc'));
    expect(trigger).toHaveAttribute('aria-expanded', 'false');
    await userEvent.click(trigger);
    await userEvent.click(screen.getByTestId('release-date-asc'));
    expect(trigger).toHaveAttribute('aria-expanded', 'false');
    await userEvent.click(trigger);
    await userEvent.click(screen.getByTestId('a-z'));
    expect(trigger).toHaveAttribute('aria-expanded', 'false');
    await userEvent.click(trigger);
    await userEvent.click(screen.getByTestId('z-a'));
    expect(trigger).toHaveAttribute('aria-expanded', 'false');
    await userEvent.click(trigger);
    await userEvent.click(screen.getByTestId('metascore'));
    expect(trigger).toHaveAttribute('aria-expanded', 'false');
    await userEvent.click(trigger);
    await userEvent.click(screen.getByTestId('user-rating'));
    expect(trigger).toHaveAttribute('aria-expanded', 'false');
  });
});
