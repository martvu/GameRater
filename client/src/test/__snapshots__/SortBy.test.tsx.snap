// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SortBy Component > matches the snapshot 1`] = `
<DocumentFragment>
  <button
    aria-autocomplete="none"
    aria-controls="radix-:r9:"
    aria-expanded="false"
    aria-label="Sort By Select"
    class="flex h-10 items-center justify-between rounded-md border border-input bg-background px-3 py-2 text-sm ring-offset-background placeholder:text-muted-foreground focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 w-[150px] hover:bg-accent"
    data-placeholder=""
    data-state="closed"
    data-testid="sort-by-select"
    dir="ltr"
    role="combobox"
    type="button"
  >
    <span
      style="pointer-events: none;"
    >
      Sort By
    </span>
    <svg
      aria-hidden="true"
      class="h-4 w-4 opacity-50"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m6 9 6 6 6-6"
      />
    </svg>
  </button>
</DocumentFragment>
`;
